### PR TITLE
Memory WIP

### PR DESCRIPTION
**Pull request by MrChico**
This is not in any finished state, but I'd appreaciate any feedback before I continue if you find some time @ehildenb

### CODE REVIEW

**Comment by ehildenb** [comment]
@MrChico did you manage to get it to build on your system? What changes were needed? If so, we should get #22 fixed and merged so we can look at these changes in isolation.

**Comment by ehildenb** [comment]
Generally GitHub does not behave well when stacking PRs, which can lead to lots of confusion, and the temptation to clone/re-open PRs over and over again. Instead I try to only have one PR open/actively worked on by me at a time.

File: test.md [hunk 1]
```diff
@@ -76,5 +76,22 @@ This simply checks that the given function exists in the `<funcs>` cell and has
...
          </funcs>
+```
+
+### Memory Assertions
+
+Checks that memory is in a certain state
+
+```k
+    syntax Instr ::= "#assertMemory" Int Int Bytes String
+ // -----------------------------------------------------------------------
```
Comment (ehildenb) [inline test.md]: formatting, `----------` underneed should line up with end of line

File: test.md [hunk 2]
```diff
@@ -76,5 +76,22 @@ This simply checks that the given function exists in the `<funcs>` cell and has
...
+    rule <k> #assertMemory MIN MAX DATA DESCRIPTION => . ... </k>
+         <mems>
+           <memInst>
+              <min> MIN </min>
+              <max> MAX </max>
+              <content> DATA </content>
+           </memInst> => .Bag
```
Comment (ehildenb) [inline test.md]: I would prefer formatting more like this:

```
         ( <memInst>
              <min> MIN </min>
              <max> MAX </max>
              <content> DATA </content>
           </memInst>
        => .Bag
         )
```

Also, do you want a `...` inside the `<mems>` to capture any additional `<memInst>` that may exist?

File: tests/simple/memory.wast [hunk 3]
```diff
@@ -1,33 +1,62 @@
...
+  #assertTopStack < i32 > 77 "M in bytes" 
+)
+
+(module
+  (memory 1)
+  (data (i32.const 0) "WASM")
+  (i32.load8_u (i32.const 4))
+  #assertTopStack < i32 > 0 "0 in bytes" 
+)
+;)
```
Comment (ehildenb) [inline tests/simple/memory.wast]: I would prefer to be able to test memories without having them wrapped in a module if possible. This makes the language more "incremental", in the sense that someone can just pick it up and start writing as if at a repl without having to load up a module.

These hand-written tests should be about these atomic incremental actions that can be taken, whereas the actual wasm test-set can be used for full-blown module-based tests.

**Comment by MrChico** [comment]
> @MrChico did you manage to get it to build on your system? What changes were needed? If so, we should get #22 fixed and merged so we can look at these changes in isolation.

It builds, but doesn't run with OCaml, no. Still have that linking problem:
```
error loading shared library: dlopen(/Users/martinlundfall/wasm-semantics/.build/defn/test/test-kompiled/realdef.cmxs, 10): Symbol not found: _caml_curry10
  Referenced from: /Users/martinlundfall/wasm-semantics/.build/defn/test/test-kompiled/realdef.cmxs
  Expected in: flat namespace
```
the reference is undefined, as also shown by
```
nm .build/defn/wasm/wasm-kompiled/realdef.cmxs | egrep '_caml_curry'
                 U _caml_curry10
                 U _caml_curry11
0000000000000e60 t _caml_curry2
0000000000000eb0 t _caml_curry2_1
0000000000000d70 t _caml_curry3
0000000000000df0 t _caml_curry3_1
...
```

so I'm working with the java backend for now.

File: test.md (hunk 2 above)
Comment (MrChico) [reply test.md]: The current version of WASM seems to only support one memory instance. So either there is no memory at all, or there is a memory instance. I reflected this in the configuration as:
`<memInst multiplicity="?">   //current WASM supports at most 1 memory instance`.

File: tests/simple/memory.wast (hunk 3 above)
Comment (MrChico) [reply tests/simple/memory.wast]: Good point I'll fix that

File: test.md (hunk 2 above)
Comment (ehildenb) [reply test.md]: Hmmmm, in that case it makes sense. Then we can just switch `multiplicity="*"` to get multiple memory instance support. Please make the formatting change as well.

File: data.md [hunk 4]
```diff
@@ -185,5 +187,70 @@ Operator `_++_` implements an append operator for sort `Stack`.
 
     rule #drop(.ValTypes,   STACK)                       => STACK
     rule #drop(TYPE VTYPES, < TYPE > VAL:Number : STACK) => #drop(VTYPES, STACK)
+```
+Byte Map
+--------
```
Comment (hjorthjort) [inline data.md]: @ehildenb  I want to integrate this work by @MrChico. Is using the `Bytes` abstractions a good way to go, or will it be hard to use for proofs? We could also use `Int` and build some custom abstractions.

**Comment by ehildenb** [comment]
Closed for #24.

**Pull request closed without merging**